L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc

```diff
@@ -0,0 +1,327 @@
+/*!  \brief   Implementation of methods of TTClusterBuilder.h
+ *  \details Here, in the source file, the methods which do depend
+ *           on the specific type <T> that can fit the template.
+ *
+ * \author Andrew W. Rose
+ * \author Nicola Pozzobon
+ * \author Ivan Reid
+ * \date 2013, Jul 18
+ *
+ */
+
+#include "L1Trigger/TrackTrigger/plugins/TTStubBuilder.h"
+
+/// Implement the producer
+template< >
+void TTStubBuilder< Ref_Phase2TrackerDigi_ >::produce( edm::Event& iEvent, const edm::EventSetup& iSetup )
+{
+  //Retrieve tracker topology from geometry                                                                                                              
+  edm::ESHandle<TrackerTopology> tTopoHandle;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoHandle);
+  const TrackerTopology* const tTopo = tTopoHandle.product();
+  edm::ESHandle< TrackerGeometry > tGeomHandle;
+  iSetup.get< TrackerDigiGeometryRecord >().get( tGeomHandle );
+  const TrackerGeometry* const theTrackerGeom = tGeomHandle.product();
+	
+  /// Prepare output
+  std::auto_ptr< edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > > TTClusterDSVForOutput( new edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputTemp( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputAccepted( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputRejected( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+
+  /// Get the Clusters already stored away
+  edm::Handle< edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > > clusterHandle;
+  iEvent.getByToken( clustersToken, clusterHandle );
+
+  /// Get the maximum number of stubs per ROC
+  /// (CBC3-style)
+  //  unsigned maxStubs = theStackedTracker->getCBC3MaxStubs();
+  unsigned maxStubs = 3;
+
+  for (auto gd=theTrackerGeom->dets().begin(); gd != theTrackerGeom->dets().end(); gd++) {
+      DetId detid = (*gd)->geographicalId();
+      if(detid.subdetId()!=StripSubdetector::TOB && detid.subdetId()!=StripSubdetector::TID ) continue; // only run on OT
```

Comment:
I am afraid that if we change geometry definition for some reason, this is not working anymore. Can you maybe think something else? look in my private branch [here](https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...ebrondol:phase2pixel_geometry_81X#diff-9190c25b9b1ca19539802a4ecea6f6cbR155) and check if this can be a better idea..